lib/ABI/Definition.cpp

```diff
@@ -0,0 +1,403 @@
+  }
+
+  // Make sure the architecture of of the register is as expected.
+  const auto Architecture = model::ABI::getRegisterArchitecture(D.ABI());
+  const model::Register::Values RVLR = D.ReturnValueLocationRegister();
+  if (model::Register::getArchitecture(RVLR) != Architecture)
+    return false;
+
+  if (isVectorRegister(D.ReturnValueLocationRegister())) {
+    // Vector register used as the return value locations are not supported.
+    return false;
+  } else if (llvm::is_contained(D.CalleeSavedRegisters(),
+                                D.ReturnValueLocationRegister())) {
+    // Using callee saved register as a return value location doesn't make
+    // much sense: filter those out.
+    return false;
+  } else {
+    // The return value location register can optionally also be the first
+    // GPRs, but only the first one.
+    const auto &GPRs = D.GeneralPurposeArgumentRegisters();
+    const auto CacheIterator = llvm::find(GPRs,
+                                          D.ReturnValueLocationRegister());
+    if (CacheIterator != GPRs.end() && CacheIterator != GPRs.begin())
+      return false;
+  }
+
+  return true;
+}
+
+using RegisterVector = std::vector<model::Register::Values>;
+
+template<typename RegisterType>
+bool compatibilityHelper(const SortedVector<RegisterType> &UsedRegisters,
+                         const RegisterVector &AllowedRegisters,
+                         model::Architecture::Values Architecture) {
+  for (const RegisterType &Register : UsedRegisters) {
+    // Verify the architecture of used registers.
+    if (model::Register::getArchitecture(Register.Location()) != Architecture)
+      revng_abort();
+  }
+
+  // Verify that every used register is also allowed.
+  for (const RegisterType &Register : UsedRegisters)
+    if (llvm::count(AllowedRegisters, Register.Location()) != 1)
+      return false;
+
+  return true;
+}
+
+namespace abi {
+
+bool Definition::isValid() const {
+  if (ABI() == model::ABI::Invalid)
+    return false;
+
+  const auto Architecture = model::ABI::getRegisterArchitecture(ABI());
+  if (!verifyRegisters(GeneralPurposeArgumentRegisters(), Architecture))
+    return false;
+  if (!verifyRegisters(GeneralPurposeReturnValueRegisters(), Architecture))
+    return false;
+  if (!verifyRegisters(VectorArgumentRegisters(), Architecture))
+    return false;
+  if (!verifyRegisters(VectorReturnValueRegisters(), Architecture))
+    return false;
+  if (!verifyRegisters(CalleeSavedRegisters(), Architecture))
+    return false;
+
+  if (!verifyReturnValueLocation(*this))
+    return false;
+
+  // TODO: extend to verify `Types` once it's more substantial,
+  //       i.e. holds vector argument information.
+
+  return true;
+}
+
+template<typename... Ts>
+using SV = SortedVector<Ts...>;
+
+template<typename AR, typename RVR>
+bool Definition::checkCompatibility(const SV<AR> &ArgumentRegisters,
+                                    const SV<RVR> &ReturnValueRegisters) const {
```

Comment:
> Can you have `using ArgumentVector` or the like? SV might also stand for `SmallVector`...

This thing is a template for a reason, it get called with different `AR`s and `RVR`s (which are the names I hate with passion too), so it needs to stay a template one way or the other.

> Also elsewhere you're shortening `Definition` to `D`, which is still sad but a bit better.

I tried, but even that wasn't enough at this point: it's exactly 80 symbols as is and `SortedVector` is two symbols longer than `Definition` (so usage of `D` is exactly one symbol too long :cry:).

The only way I can go to sleep with an awful alias like that is because it's a part of a relatively small `.cpp`-file and the alias declaration is right on top of the only user.

Reply:
Closing this as it can be solved by getting rid of template parameters, see [here](https://github.com/revng/revng/pull/232#discussion_r1054743577).